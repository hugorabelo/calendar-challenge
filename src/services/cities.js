import axios from 'axios'
import httpAdapter from 'axios/lib/adapters/http'

const API_CITY_KEY = process.env.VUE_APP_API_CITY_KEY

const http = axios.create({
  baseURL: 'https://andruxnet-world-cities-v1.p.rapidapi.com/',
  adapter: httpAdapter,
  headers: {
    'X-RapidAPI-Host': 'andruxnet-world-cities-v1.p.rapidapi.com',
    'X-RapidAPI-Key': API_CITY_KEY,
  },
})

export function getCities(partial) {
  return http.get(`?query=${partial}&searchby=city`)
}
