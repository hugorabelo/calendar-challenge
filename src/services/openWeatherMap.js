import axios from 'axios'
import httpAdapter from 'axios/lib/adapters/http'

const API_KEY = '0d8aa4b5581a5331b53fe1c53c106c4f'

const http = axios.create({
  baseURL: 'https://api.openweathermap.org/data/2.5/',
  adapter: httpAdapter,
})

export async function getWeatherForecast(city) {
  let excluded = 'current,alerts,minutely,hourly'
  return await http.get(`onecall?lat=${city.coord.lat}&lon=${city.coord.lon}&exclude=${excluded}&appid=${API_KEY}`)
}

export function getWeatherForecastByGeoLocation(geoLocation) {
  let excluded = 'current,alerts,minutely,hourly'
  return http.get(`onecall?lat=${geoLocation.lat}&lon=${geoLocation.lon}&exclude=${excluded}&appid=${API_KEY}`)
}

export function getCityGeoLocation(city) {
  return http.get(`weather?q=${city}&APPID=${API_KEY}`)
}
