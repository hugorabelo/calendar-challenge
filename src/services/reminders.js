import store from '@/store'

export function getReminders() {
  return store.getters.reminders
}

export function getReminderById(id) {
  let selectedReminder = store.getters.reminders.filter((rem) => rem.id == id)
  return selectedReminder.length > 0 ? selectedReminder[0] : null
}

export function addReminder(data) {
  store.dispatch('addReminder', data)
}

export function updateReminder(data) {
  store.dispatch('updateReminder', data)
}

export function deleteReminder(data) {
  store.dispatch('deleteReminder', data)
}

export function deleteRemindersGroup(data) {
  store.dispatch('deleteRemindersGroup', data)
}
