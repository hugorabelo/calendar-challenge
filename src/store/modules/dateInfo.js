import { DateTime } from 'luxon'

const state = {
  selectedMonth: DateTime.now().startOf('month'),
}
const getters = {
  selectedMonth: (state) => state.selectedMonth,
}
const mutations = {
  setSelectedMonth(state, data) {
    state.selectedMonth = data
  },
}
const actions = {
  initSelectedMonth({ commit }) {
    commit('setSelectedMonth', DateTime.now().startOf('month'))
  },
  goToNextMonth({ commit, state, dispatch }) {
    if (!state.selectedMonth) {
      dispatch('initSelectedMonth')
    }
    let nextMonth = state.selectedMonth.plus({ months: 1 })
    commit('setSelectedMonth', nextMonth)
  },
  goToPreviousMonth({ commit, state, dispatch }) {
    if (!state.selectedMonth) {
      dispatch('initSelectedMonth')
    }
    let previousMonth = state.selectedMonth.minus({ months: 1 })
    commit('setSelectedMonth', previousMonth)
  },
}

export default {
  state,
  getters,
  actions,
  mutations,
}
