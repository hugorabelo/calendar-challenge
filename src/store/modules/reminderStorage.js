const state = {
  reminders: localStorage.getItem('reminders') != null ? JSON.parse(localStorage.getItem('reminders')) : [],
}
const getters = {
  reminders: (state) => state.reminders,
}
const mutations = {
  setReminders(state, data) {
    state.reminders = data
    localStorage.setItem('reminders', JSON.stringify(state.reminders))
  },
}
const actions = {
  addReminder({ commit, state }, data) {
    if (state.reminders.filter((rem) => rem.id == data.id).length == 0) {
      state.reminders.push(data)
    }
    commit('setReminders', state.reminders)
  },
  updateReminder({ commit, state }, data) {
    for (var i = 0; i < state.reminders.length; i++) {
      if (state.reminders[i].id == data.id) {
        state.reminders[i] = data
      }
    }
    commit('setReminders', state.reminders)
  },
  deleteReminder({ commit, state }, data) {
    for (var i = 0; i < state.reminders.length; i++) {
      if (state.reminders[i].id == data.id) {
        state.reminders.splice(i, 1)
      }
    }
    commit('setReminders', state.reminders)
  },
  deleteRemindersGroup({ dispatch }, data) {
    for (var i = 0; i < data.length; i++) {
      dispatch('deleteReminder', { id: data[i].id })
    }
  },
}

export default {
  state,
  getters,
  actions,
  mutations,
}
