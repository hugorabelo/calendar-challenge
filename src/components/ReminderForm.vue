<template>
  <div class="">
    <div class="mb-6">
      <label for="input-description" class="text-left block mb-2 text-sm font-medium text-gray-900 dark:text-gray-300"
        >What?</label
      >
      <div class="relative">
        <div class="flex absolute inset-y-0 left-0 items-center pl-3 pointer-events-none">
          <font-awesome-icon icon="fa-solid fa-note-sticky" class="w-5 h-5 text-gray-500" />
        </div>
        <input
          id="input-description"
          class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full pl-10 p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
          v-model="formData.description"
          maxlength="30"
        />
      </div>
    </div>
    <div class="mb-6">
      <label for="input-datetime" class="text-left block mb-2 text-sm font-medium text-gray-900 dark:text-gray-300"
        >When?</label
      >
      <div class="relative">
        <div class="flex absolute inset-y-0 left-0 items-center pl-3 pointer-events-none">
          <font-awesome-icon icon="fa-solid fa-calendar" class="w-5 h-5 text-gray-500" />
        </div>
        <div
          class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full pl-10 p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
        >
          <vue-date-time input-id="input-datetime" v-model="dateTimeISO" use12-hour type="datetime"></vue-date-time>
        </div>
      </div>
    </div>
    <div class="mb-6">
      <label for="input-city" class="text-left block mb-2 text-sm font-medium text-gray-900 dark:text-gray-300"
        >Where?</label
      >
      <div class="relative">
        <div class="flex absolute inset-y-0 left-0 items-center pl-3 pointer-events-none">
          <font-awesome-icon icon="fa-solid fa-location-dot" class="w-5 h-5 text-gray-500" />
        </div>
        <select
          id="input-city"
          class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 pl-10 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
          v-model="formData.city"
        >
          <option v-for="(city, index) of cities" :key="index">{{ city }}</option>
        </select>
      </div>
    </div>
    <div class="mb-3 space-y-2 w-full text-xs">
      <label for="input-color" class="flex font-semibold text-gray-800 text-left">Please select a color</label>
      <color-picker id="input-color" v-model="formData.color"></color-picker>
    </div>
    <div class="flex justify-right">
      <button
        id="btn-cancel"
        class="mx-auto py-2.5 px-5 mr-2 mb-2 text-sm font-medium text-gray-900 focus:outline-none bg-white rounded-lg border border-gray-200 hover:bg-gray-100 hover:text-blue-700 focus:z-10 focus:ring-4 focus:ring-gray-200 dark:focus:ring-gray-700 dark:bg-gray-800 dark:text-gray-400 dark:border-gray-600 dark:hover:text-white dark:hover:bg-gray-700"
        @click="cancelEdit()"
      >
        Cancel
      </button>
      <button
        id="btn-save"
        class="text-white bg-blue-700 hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 font-medium rounded-lg text-sm px-5 py-2.5 mr-2 mb-2 dark:bg-blue-600 dark:hover:bg-blue-700 focus:outline-none dark:focus:ring-blue-800"
        @click="save()"
      >
        Save
      </button>
    </div>
    <div v-if="formData.color" class="mt-2 h-7 rounded-lg border" :class="formData.color"></div>
  </div>
</template>

<script>
import ColorPicker from './ColorPicker.vue'
import * as reminderApi from '@/services/reminders'
import * as weatherApi from '@/services/openWeatherMap'
import dateTimeMixins from '@/mixins/dateTimeMixins'
import { Datetime as VueDateTime } from 'vue-datetime'

export default {
  name: 'ReminderForm',
  components: {
    ColorPicker,
    VueDateTime,
  },
  mixins: [dateTimeMixins],
  props: {
    editingReminder: {
      type: Object,
      default: function () {
        return {
          id: null,
          description: null,
          dateTime: null,
          city: null,
          color: null,
          weatherForecast: null,
          weatherIcon: null,
        }
      },
    },
  },
  data() {
    return {
      formData: {
        ...this.editingReminder,
      },
      dateTimeISO: this.getISODateTime(this.editingReminder.dateTime),
      cities: [
        'Tokyo',
        'Jakarta',
        'Delhi',
        'Seoul',
        'Shanghai',
        'Beijing',
        'New York',
        'Sao Paulo',
        'Mexico City',
        'Los Angeles',
        'Buenos Aires',
        'Istanbul',
        'Rio de Janeiro',
        'Paris',
        'London',
      ],
    }
  },
  computed: {
    isEdit() {
      return this.formData.id !== null
    },
  },
  methods: {
    async save() {
      let { data } = await weatherApi.getWeatherForecast(this.formData.city)
      this.formData.weatherForecast = ''
      this.formData.weatherIcon = ''
      let countDays = await this.countDaysFromNow(this.dateTimeISO)
      if (countDays <= data.daily.length) {
        this.formData.weatherForecast = data.daily[countDays].weather[0].description
        this.formData.weatherIcon = data.daily[countDays].weather[0].icon
      }
      this.formData.dateTime = this.dateTimeISO
      if (this.formData.description.length > 30) {
        this.formData.description = this.formData.description.substring(0, 30)
      }
      if (this.isEdit) {
        await reminderApi.updateReminder(this.formData)
      } else {
        this.formData.id = this.guid()
        await reminderApi.addReminder(this.formData)
      }
      let currentReminder = {
        ...this.formData,
        dateTime: this.getObjectDateTime(this.formData.dateTime),
      }
      this.$emit('save', currentReminder)
    },
    async cancelEdit() {
      this.$emit('cancel')
    },
    guid() {
      return (
        this.s4() +
        this.s4() +
        '-' +
        this.s4() +
        '-' +
        this.s4() +
        '-' +
        this.s4() +
        '-' +
        this.s4() +
        this.s4() +
        this.s4()
      )
    },
    s4() {
      return Math.floor((1 + Math.random()) * 0x10000)
        .toString(16)
        .substring(1)
    },
  },
}
</script>

<style></style>
