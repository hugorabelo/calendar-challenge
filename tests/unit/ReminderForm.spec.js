import { mount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import flushPromisses from 'flush-promises'
import nock from 'nock'
import * as weatherApi from '@/services/openWeatherMap'
import ReminderForm from '@/components/ReminderForm'
import { DateTime } from 'luxon'
import dateInfo from '@/store/modules/dateInfo'
import reminderStorage from '@/store/modules/reminderStorage'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
import { library } from '@fortawesome/fontawesome-svg-core'
import {
  faBan,
  faEdit,
  faTrash,
  faXmark,
  faNoteSticky,
  faCalendar,
  faLocationDot,
} from '@fortawesome/free-solid-svg-icons'
import weatherData from './fixtures/weatherData'
import cityData from './fixtures/cityData'

const localVue = createLocalVue()
localVue.use(Vuex)
library.add(faBan, faEdit, faTrash, faXmark, faNoteSticky, faCalendar, faLocationDot)
localVue.component('font-awesome-icon', FontAwesomeIcon)

describe('ReminderForm', () => {
  const spy = jest.fn()

  const expectedDataCity = {
    q: 'London',
    APPID: '0d8aa4b5581a5331b53fe1c53c106c4f',
  }

  const expectedDataWeather = {
    lat: '51.5085',
    lon: '-0.1257',
    exclude: 'current,alerts,minutely,hourly',
    appid: '0d8aa4b5581a5331b53fe1c53c106c4f',
  }

  const reminder = {
    id: '123-456',
    description: 'Event 01',
    dateTime: DateTime.now().toISO(),
    city: 'London',
    color: 'bg-red-600',
    weatherForecast: 'light rain',
    weatherIcon: '10n',
  }

  const build = (options) => {
    return mount(ReminderForm, {
      localVue,
      store: new Vuex.Store({
        modules: {
          dateInfo: { ...dateInfo },
          reminderStorage: { ...reminderStorage },
        },
      }),
      ...options,
      destroyed() {
        spy()
      },
    })
  }

  test('Is a Vue instance', () => {
    const { vm } = build({
      propsData: {
        reminderData: reminder,
      },
    })
    expect(vm).toBeTruthy()
  })

  test('Save new reminder', async () => {
    const wrapper = build({
      propsData: { dateTimeISO: reminder.dateTime },
    })

    const requestCity = nock('https://api.openweathermap.org/data/2.5')
      .persist()
      .get('/weather')
      .query(expectedDataCity)
      .reply(200, cityData)

    const requestWeather = nock('https://api.openweathermap.org/data/2.5')
      .persist()
      .get('/onecall')
      .query(expectedDataWeather)
      .reply(200, weatherData)

    const responseCity = await weatherApi.getCityGeoLocation(reminder.city)
    await flushPromisses()

    const responseWeather = await weatherApi.getWeatherForecast(cityData)
    await flushPromisses()

    expect(responseCity.data).toEqual(cityData)
    expect(requestCity.isDone()).toBe(true)

    expect(responseWeather.data).toEqual(weatherData)
    expect(requestWeather.isDone()).toBe(true)

    expect(wrapper.html()).toMatchSnapshot()

    let inputDescription = wrapper.findComponent('#input-description')
    let options = wrapper.findComponent('#input-city').findAll('option')
    let btnSave = wrapper.findComponent('#btn-save')

    await inputDescription.setValue(reminder.description)
    await options.at(-1).setSelected()

    await btnSave.trigger('click')
    expect(wrapper.emitted('save')).toBeTruthy()
  })
})
