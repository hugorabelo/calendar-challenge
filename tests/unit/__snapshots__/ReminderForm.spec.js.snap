// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReminderForm Save new reminder 1`] = `
<div datetimeiso="2022-04-01T00:12:45.322-03:00">
  <div class="mb-6"><label for="input-description" class="text-left block mb-2 text-sm font-medium text-gray-900 dark:text-gray-300">What?</label>
    <div class="relative">
      <div class="flex absolute inset-y-0 left-0 items-center pl-3 pointer-events-none"><svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="note-sticky" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512" class="w-5 h-5 text-gray-500 svg-inline--fa fa-note-sticky">
          <path fill="currentColor" d="M400 32h-352C21.49 32 0 53.49 0 80v352C0 458.5 21.49 480 48 480h245.5c16.97 0 33.25-6.744 45.26-18.75l90.51-90.51C441.3 358.7 448 342.5 448 325.5V80C448 53.49 426.5 32 400 32zM64 96h320l-.001 224H320c-17.67 0-32 14.33-32 32v64H64V96z" class=""></path>
        </svg></div> <input id="input-description" maxlength="30" class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full pl-10 p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500">
    </div>
  </div>
  <div class="mb-6"><label for="input-datetime" class="text-left block mb-2 text-sm font-medium text-gray-900 dark:text-gray-300">When?</label>
    <div class="relative">
      <div class="flex absolute inset-y-0 left-0 items-center pl-3 pointer-events-none"><svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="calendar" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512" class="w-5 h-5 text-gray-500 svg-inline--fa fa-calendar">
          <path fill="currentColor" d="M96 32C96 14.33 110.3 0 128 0C145.7 0 160 14.33 160 32V64H288V32C288 14.33 302.3 0 320 0C337.7 0 352 14.33 352 32V64H400C426.5 64 448 85.49 448 112V160H0V112C0 85.49 21.49 64 48 64H96V32zM448 464C448 490.5 426.5 512 400 512H48C21.49 512 0 490.5 0 464V192H448V464z" class=""></path>
        </svg></div>
      <div class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full pl-10 p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500">
        <div class="vdatetime"> <input id="input-datetime" type="text" class="vdatetime-input">
          <!---->
          <transition-group-stub tag="div" name="vdatetime-fade">
            <!---->
            <!---->
          </transition-group-stub>
        </div>
      </div>
    </div>
  </div>
  <div class="mb-6"><label for="input-city" class="text-left block mb-2 text-sm font-medium text-gray-900 dark:text-gray-300">Where?</label>
    <div class="relative">
      <div class="flex absolute inset-y-0 left-0 items-center pl-3 pointer-events-none"><svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="location-dot" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 384 512" class="w-5 h-5 text-gray-500 svg-inline--fa fa-location-dot">
          <path fill="currentColor" d="M168.3 499.2C116.1 435 0 279.4 0 192C0 85.96 85.96 0 192 0C298 0 384 85.96 384 192C384 279.4 267 435 215.7 499.2C203.4 514.5 180.6 514.5 168.3 499.2H168.3zM192 256C227.3 256 256 227.3 256 192C256 156.7 227.3 128 192 128C156.7 128 128 156.7 128 192C128 227.3 156.7 256 192 256z" class=""></path>
        </svg></div> <select id="input-city" class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 pl-10 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500">
        <option>Tokyo</option>
        <option>Jakarta</option>
        <option>Delhi</option>
        <option>Seoul</option>
        <option>Shanghai</option>
        <option>Beijing</option>
        <option>New York</option>
        <option>Sao Paulo</option>
        <option>Mexico City</option>
        <option>Los Angeles</option>
        <option>Buenos Aires</option>
        <option>Istanbul</option>
        <option>Rio de Janeiro</option>
        <option>Paris</option>
        <option>London</option>
      </select>
    </div>
  </div>
  <div class="mb-3 space-y-2 w-full text-xs"><label for="input-color" class="flex font-semibold text-gray-800 text-left">Please select a color</label>
    <div class="flex" id="input-color">
      <div class="flex h-8 w-8 mr-2 rounded cursor-pointer bg-red-500"></div>
      <div class="flex h-8 w-8 mr-2 rounded cursor-pointer bg-yellow-500"></div>
      <div class="flex h-8 w-8 mr-2 rounded cursor-pointer bg-green-500"></div>
      <div class="flex h-8 w-8 mr-2 rounded cursor-pointer bg-blue-500"></div>
      <div class="flex h-8 w-8 mr-2 rounded cursor-pointer bg-purple-500"></div>
      <div class="flex h-8 w-8 mr-2 rounded cursor-pointer bg-pink-500"></div>
    </div>
  </div>
  <div class="flex justify-right"><button id="btn-cancel" class="mx-auto py-2.5 px-5 mr-2 mb-2 text-sm font-medium text-gray-900 focus:outline-none bg-white rounded-lg border border-gray-200 hover:bg-gray-100 hover:text-blue-700 focus:z-10 focus:ring-4 focus:ring-gray-200 dark:focus:ring-gray-700 dark:bg-gray-800 dark:text-gray-400 dark:border-gray-600 dark:hover:text-white dark:hover:bg-gray-700">
      Cancel
    </button> <button id="btn-save" class="text-white bg-blue-700 hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 font-medium rounded-lg text-sm px-5 py-2.5 mr-2 mb-2 dark:bg-blue-600 dark:hover:bg-blue-700 focus:outline-none dark:focus:ring-blue-800">
      Save
    </button></div>
  <!---->
</div>
`;
